### ISSUE TITLE
500 error when attempting to list uninitialized sleds

### ISSUE THREAD

**Issue by askfongjojo** [oxidecomputer/omicron#4621]
I ran into this on rack2 when invoking the Nexus API for getting a list of uninitialized sleds:
```
00:32:34.861Z INFO 65a11c18-7f59-41ac-b9e7-680627f996e7 (dropshot_external): request completed
    error_message_external = Internal Server Error
    error_message_internal = hit limit of 50 records while loading collection
    file = /home/build/.cargo/git/checkouts/dropshot-a4a923d29dccc492/ff87a01/dropshot/src/server.rs:841
    latency_us = 67030
    local_addr = 172.30.2.5:443
    method = GET
    remote_addr = 172.20.17.42:53378
    req_id = 61f13361-3699-4d64-9af9-308ded641a60
    response_code = 500
    uri = /v1/system/hardware/uninitialized-sleds
```

I'm not entirely sure what the offending table is (definitely not the sled table which has fewer than 60 rows). This is the content of the `inv_service_processor` table
```
root@[fd00:1122:3344:105::3]:32221/omicron> select * from inv_service_processor order by sp_type, sp_slot;
           inv_collection_id           |           hw_baseboard_id            |        time_collected         |                source                | sp_type | sp_slot | baseboard_revision | hubris_archive_id | power_state
---------------------------------------+--------------------------------------+-------------------------------+--------------------------------------+---------+---------+--------------------+-------------------+--------------
  a15e1a80-d9aa-4679-ba21-896ba86d9de8 | c6313404-bde5-42f8-a884-9de41c06843f | 2023-11-30 18:51:35.736285+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |       8 |                  6 | 62882ff9ad414aae  | A0
  5247e759-2478-4be8-b8ed-795e1d2ed15c | c6313404-bde5-42f8-a884-9de41c06843f | 2023-12-06 01:00:25.684695+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |       8 |                  6 | 62882ff9ad414aae  | A0
  68b2d1c0-248d-4f00-9a20-a3153a20c47f | c6313404-bde5-42f8-a884-9de41c06843f | 2023-12-06 00:59:43.813077+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |       8 |                  6 | 62882ff9ad414aae  | A0
  4e31b5ee-f2b9-4288-bcea-34a37a4f2a2e | c6313404-bde5-42f8-a884-9de41c06843f | 2023-12-06 00:59:43.236164+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |       8 |                  6 | 62882ff9ad414aae  | A0
  68b2d1c0-248d-4f00-9a20-a3153a20c47f | ede87635-3fa6-4957-b109-2309841f55f5 | 2023-12-06 00:59:48.77332+00  | http://[fd00:1122:3344:10b::2]:12225 | sled    |       9 |                  6 | 62882ff9ad414aae  | A0
  a15e1a80-d9aa-4679-ba21-896ba86d9de8 | ede87635-3fa6-4957-b109-2309841f55f5 | 2023-11-30 18:51:40.411397+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |       9 |                  6 | 62882ff9ad414aae  | A0
  4e31b5ee-f2b9-4288-bcea-34a37a4f2a2e | ede87635-3fa6-4957-b109-2309841f55f5 | 2023-12-06 00:59:47.691355+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |       9 |                  6 | 62882ff9ad414aae  | A0
  5247e759-2478-4be8-b8ed-795e1d2ed15c | ede87635-3fa6-4957-b109-2309841f55f5 | 2023-12-06 01:00:30.42979+00  | http://[fd00:1122:3344:10b::2]:12225 | sled    |       9 |                  6 | 62882ff9ad414aae  | A0
  a15e1a80-d9aa-4679-ba21-896ba86d9de8 | 0d40687a-8048-4a03-8fd5-8924478576b0 | 2023-11-30 18:51:36.249766+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      10 |                  6 | 62882ff9ad414aae  | A0
  4e31b5ee-f2b9-4288-bcea-34a37a4f2a2e | 0d40687a-8048-4a03-8fd5-8924478576b0 | 2023-12-06 00:59:43.809217+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      10 |                  6 | 62882ff9ad414aae  | A0
  5247e759-2478-4be8-b8ed-795e1d2ed15c | 0d40687a-8048-4a03-8fd5-8924478576b0 | 2023-12-06 01:00:26.676803+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      10 |                  6 | 62882ff9ad414aae  | A0
  68b2d1c0-248d-4f00-9a20-a3153a20c47f | 0d40687a-8048-4a03-8fd5-8924478576b0 | 2023-12-06 00:59:44.390158+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      10 |                  6 | 62882ff9ad414aae  | A0
  68b2d1c0-248d-4f00-9a20-a3153a20c47f | 58338408-3428-40af-8e91-522583eabbde | 2023-12-06 00:59:49.35394+00  | http://[fd00:1122:3344:10b::2]:12225 | sled    |      11 |                  6 | 62882ff9ad414aae  | A0
  4e31b5ee-f2b9-4288-bcea-34a37a4f2a2e | 58338408-3428-40af-8e91-522583eabbde | 2023-12-06 00:59:48.267075+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      11 |                  6 | 62882ff9ad414aae  | A0
  5247e759-2478-4be8-b8ed-795e1d2ed15c | 58338408-3428-40af-8e91-522583eabbde | 2023-12-06 01:00:31.007642+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      11 |                  6 | 62882ff9ad414aae  | A0
  a15e1a80-d9aa-4679-ba21-896ba86d9de8 | 58338408-3428-40af-8e91-522583eabbde | 2023-11-30 18:51:40.925884+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      11 |                  6 | 62882ff9ad414aae  | A0
  a15e1a80-d9aa-4679-ba21-896ba86d9de8 | 60ab92e2-5fdc-45f7-bbe4-3ccaef6c076c | 2023-11-30 18:51:35.225004+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      12 |                  6 | 62882ff9ad414aae  | A0
  68b2d1c0-248d-4f00-9a20-a3153a20c47f | 60ab92e2-5fdc-45f7-bbe4-3ccaef6c076c | 2023-12-06 00:59:43.236836+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      12 |                  6 | 62882ff9ad414aae  | A0
  4e31b5ee-f2b9-4288-bcea-34a37a4f2a2e | 60ab92e2-5fdc-45f7-bbe4-3ccaef6c076c | 2023-12-06 00:59:42.658833+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      12 |                  6 | 62882ff9ad414aae  | A0
  5247e759-2478-4be8-b8ed-795e1d2ed15c | 60ab92e2-5fdc-45f7-bbe4-3ccaef6c076c | 2023-12-06 01:00:24.771318+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      12 |                  6 | 62882ff9ad414aae  | A0
  a15e1a80-d9aa-4679-ba21-896ba86d9de8 | 199c809a-2cde-4864-b1b4-31066e92fccd | 2023-11-30 18:51:39.987743+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      13 |                  6 | 8eaae7036b8f25ca  | A0
  68b2d1c0-248d-4f00-9a20-a3153a20c47f | a69500c7-8753-4de6-8498-2c213363aba4 | 2023-12-06 00:59:42.659137+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      14 |                  6 | 62882ff9ad414aae  | A0
  a15e1a80-d9aa-4679-ba21-896ba86d9de8 | a69500c7-8753-4de6-8498-2c213363aba4 | 2023-11-30 18:51:34.711885+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      14 |                  6 | 62882ff9ad414aae  | A0
  5247e759-2478-4be8-b8ed-795e1d2ed15c | a69500c7-8753-4de6-8498-2c213363aba4 | 2023-12-06 01:00:24.105964+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      14 |                  6 | 62882ff9ad414aae  | A0
  4e31b5ee-f2b9-4288-bcea-34a37a4f2a2e | a69500c7-8753-4de6-8498-2c213363aba4 | 2023-12-06 00:59:42.082195+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      14 |                  6 | 62882ff9ad414aae  | A0
  68b2d1c0-248d-4f00-9a20-a3153a20c47f | 04cf6209-2071-46ff-a10e-026751ada85a | 2023-12-06 00:59:41.754155+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      16 |                  6 | 62882ff9ad414aae  | A0
  5247e759-2478-4be8-b8ed-795e1d2ed15c | 04cf6209-2071-46ff-a10e-026751ada85a | 2023-12-06 01:00:23.525164+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      16 |                  6 | 62882ff9ad414aae  | A0
  a15e1a80-d9aa-4679-ba21-896ba86d9de8 | 04cf6209-2071-46ff-a10e-026751ada85a | 2023-11-30 18:51:34.200395+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      16 |                  6 | 62882ff9ad414aae  | A0
  4e31b5ee-f2b9-4288-bcea-34a37a4f2a2e | 04cf6209-2071-46ff-a10e-026751ada85a | 2023-12-06 00:59:41.493225+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      16 |                  6 | 62882ff9ad414aae  | A0
  4e31b5ee-f2b9-4288-bcea-34a37a4f2a2e | 7a7ecd12-f4c1-4fa6-970e-45e79c972854 | 2023-12-06 00:59:45.458932+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      17 |                  6 | 62882ff9ad414aae  | A0
  68b2d1c0-248d-4f00-9a20-a3153a20c47f | 7a7ecd12-f4c1-4fa6-970e-45e79c972854 | 2023-12-06 00:59:46.618853+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      17 |                  6 | 62882ff9ad414aae  | A0
  5247e759-2478-4be8-b8ed-795e1d2ed15c | 7a7ecd12-f4c1-4fa6-970e-45e79c972854 | 2023-12-06 01:00:28.267332+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      17 |                  6 | 62882ff9ad414aae  | A0
  a15e1a80-d9aa-4679-ba21-896ba86d9de8 | 7a7ecd12-f4c1-4fa6-970e-45e79c972854 | 2023-11-30 18:51:38.019738+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      17 |                  6 | 62882ff9ad414aae  | A0
  4e31b5ee-f2b9-4288-bcea-34a37a4f2a2e | 5bc79ceb-dd38-4bfc-b44f-04bf7f6a2d6a | 2023-12-06 00:59:46.114634+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      19 |                  6 | 4a6db7d8a103107e  | A0
  a15e1a80-d9aa-4679-ba21-896ba86d9de8 | 5bc79ceb-dd38-4bfc-b44f-04bf7f6a2d6a | 2023-11-30 18:51:38.532508+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      19 |                  6 | 4a6db7d8a103107e  | A2
  5247e759-2478-4be8-b8ed-795e1d2ed15c | 5bc79ceb-dd38-4bfc-b44f-04bf7f6a2d6a | 2023-12-06 01:00:28.844143+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      19 |                  6 | 4a6db7d8a103107e  | A0
  68b2d1c0-248d-4f00-9a20-a3153a20c47f | 5bc79ceb-dd38-4bfc-b44f-04bf7f6a2d6a | 2023-12-06 00:59:47.194593+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      19 |                  6 | 4a6db7d8a103107e  | A0
  a15e1a80-d9aa-4679-ba21-896ba86d9de8 | 88093ac8-f4cd-4b25-8f36-c3713aa52d93 | 2023-11-30 18:51:39.47115+00  | http://[fd00:1122:3344:10b::2]:12225 | sled    |      21 |                  6 | 62882ff9ad414aae  | A0
  4e31b5ee-f2b9-4288-bcea-34a37a4f2a2e | 88093ac8-f4cd-4b25-8f36-c3713aa52d93 | 2023-12-06 00:59:47.114575+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      21 |                  6 | 62882ff9ad414aae  | A0
  5247e759-2478-4be8-b8ed-795e1d2ed15c | 88093ac8-f4cd-4b25-8f36-c3713aa52d93 | 2023-12-06 01:00:29.851184+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      21 |                  6 | 62882ff9ad414aae  | A0
  68b2d1c0-248d-4f00-9a20-a3153a20c47f | 88093ac8-f4cd-4b25-8f36-c3713aa52d93 | 2023-12-06 00:59:48.198498+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      21 |                  6 | 62882ff9ad414aae  | A0
  68b2d1c0-248d-4f00-9a20-a3153a20c47f | 9587940e-760b-4fbb-ae4d-c8c711c9f561 | 2023-12-06 00:59:47.6188+00   | http://[fd00:1122:3344:10b::2]:12225 | sled    |      23 |                  6 | 62882ff9ad414aae  | A0
  4e31b5ee-f2b9-4288-bcea-34a37a4f2a2e | 9587940e-760b-4fbb-ae4d-c8c711c9f561 | 2023-12-06 00:59:46.538839+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      23 |                  6 | 62882ff9ad414aae  | A0
  5247e759-2478-4be8-b8ed-795e1d2ed15c | 9587940e-760b-4fbb-ae4d-c8c711c9f561 | 2023-12-06 01:00:29.269441+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      23 |                  6 | 62882ff9ad414aae  | A0
  a15e1a80-d9aa-4679-ba21-896ba86d9de8 | 9587940e-760b-4fbb-ae4d-c8c711c9f561 | 2023-11-30 18:51:38.955196+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      23 |                  6 | 62882ff9ad414aae  | A0
  a15e1a80-d9aa-4679-ba21-896ba86d9de8 | b352e992-2390-441d-b8eb-58005bf179c3 | 2023-11-30 18:51:36.76637+00  | http://[fd00:1122:3344:10b::2]:12225 | sled    |      25 |                  6 | 62882ff9ad414aae  | A0
  5247e759-2478-4be8-b8ed-795e1d2ed15c | b352e992-2390-441d-b8eb-58005bf179c3 | 2023-12-06 01:00:27.254298+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      25 |                  6 | 62882ff9ad414aae  | A0
  68b2d1c0-248d-4f00-9a20-a3153a20c47f | b352e992-2390-441d-b8eb-58005bf179c3 | 2023-12-06 00:59:45.078762+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      25 |                  6 | 62882ff9ad414aae  | A0
  4e31b5ee-f2b9-4288-bcea-34a37a4f2a2e | b352e992-2390-441d-b8eb-58005bf179c3 | 2023-12-06 00:59:44.384789+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      25 |                  6 | 62882ff9ad414aae  | A0
  5247e759-2478-4be8-b8ed-795e1d2ed15c | 0c78a1ed-bcf5-44e5-aab0-a6c3998fdc13 | 2023-12-06 01:00:27.833786+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      27 |                  6 | 4a6db7d8a103107e  | A0
  a15e1a80-d9aa-4679-ba21-896ba86d9de8 | 0c78a1ed-bcf5-44e5-aab0-a6c3998fdc13 | 2023-11-30 18:51:37.588976+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      27 |                  6 | 4a6db7d8a103107e  | A0
  68b2d1c0-248d-4f00-9a20-a3153a20c47f | 0c78a1ed-bcf5-44e5-aab0-a6c3998fdc13 | 2023-12-06 00:59:46.170476+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      27 |                  6 | 4a6db7d8a103107e  | A0
  4e31b5ee-f2b9-4288-bcea-34a37a4f2a2e | 0c78a1ed-bcf5-44e5-aab0-a6c3998fdc13 | 2023-12-06 00:59:45.023515+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |      27 |                  6 | 4a6db7d8a103107e  | A0
  4e31b5ee-f2b9-4288-bcea-34a37a4f2a2e | 37c56f3d-29e5-488a-a25a-9387692929fd | 2023-12-06 00:59:50.121184+00 | http://[fd00:1122:3344:108::2]:12225 | switch  |       0 |                  4 | f46719086d889f99  | A0
  a15e1a80-d9aa-4679-ba21-896ba86d9de8 | 37c56f3d-29e5-488a-a25a-9387692929fd | 2023-11-30 18:51:43.005362+00 | http://[fd00:1122:3344:108::2]:12225 | switch  |       0 |                  4 | f46719086d889f99  | A0
  68b2d1c0-248d-4f00-9a20-a3153a20c47f | 37c56f3d-29e5-488a-a25a-9387692929fd | 2023-12-06 00:59:51.189048+00 | http://[fd00:1122:3344:108::2]:12225 | switch  |       0 |                  4 | f46719086d889f99  | A0
  5247e759-2478-4be8-b8ed-795e1d2ed15c | 37c56f3d-29e5-488a-a25a-9387692929fd | 2023-12-06 01:00:32.850202+00 | http://[fd00:1122:3344:108::2]:12225 | switch  |       0 |                  4 | f46719086d889f99  | A0
  5247e759-2478-4be8-b8ed-795e1d2ed15c | 3c670d32-05a0-40e4-9981-fc655295abe8 | 2023-12-06 01:00:32.165588+00 | http://[fd00:1122:3344:10b::2]:12225 | switch  |       1 |                  4 | f46719086d889f99  | A0
  4e31b5ee-f2b9-4288-bcea-34a37a4f2a2e | 3c670d32-05a0-40e4-9981-fc655295abe8 | 2023-12-06 00:59:49.422657+00 | http://[fd00:1122:3344:10b::2]:12225 | switch  |       1 |                  4 | f46719086d889f99  | A0
  a15e1a80-d9aa-4679-ba21-896ba86d9de8 | 3c670d32-05a0-40e4-9981-fc655295abe8 | 2023-11-30 18:51:42.390342+00 | http://[fd00:1122:3344:10b::2]:12225 | switch  |       1 |                  4 | f46719086d889f99  | A0
  68b2d1c0-248d-4f00-9a20-a3153a20c47f | 3c670d32-05a0-40e4-9981-fc655295abe8 | 2023-12-06 00:59:50.513586+00 | http://[fd00:1122:3344:10b::2]:12225 | switch  |       1 |                  4 | f46719086d889f99  | A0
  a15e1a80-d9aa-4679-ba21-896ba86d9de8 | 3b63ab8a-302e-468f-aa2a-9b7db008a53b | 2023-11-30 18:51:41.484837+00 | http://[fd00:1122:3344:10b::2]:12225 | power   |       0 |                  8 | 2f1061bef02f947b  | A2
  68b2d1c0-248d-4f00-9a20-a3153a20c47f | 3b63ab8a-302e-468f-aa2a-9b7db008a53b | 2023-12-06 00:59:49.929013+00 | http://[fd00:1122:3344:10b::2]:12225 | power   |       0 |                  8 | 2f1061bef02f947b  | A2
  5247e759-2478-4be8-b8ed-795e1d2ed15c | 3b63ab8a-302e-468f-aa2a-9b7db008a53b | 2023-12-06 01:00:31.58716+00  | http://[fd00:1122:3344:10b::2]:12225 | power   |       0 |                  8 | 2f1061bef02f947b  | A2
  4e31b5ee-f2b9-4288-bcea-34a37a4f2a2e | 3b63ab8a-302e-468f-aa2a-9b7db008a53b | 2023-12-06 00:59:48.843075+00 | http://[fd00:1122:3344:10b::2]:12225 | power   |       0 |                  8 | 2f1061bef02f947b  | A2
(65 rows)
```
Many of the sleds have 4 entries. The number of rows with unique `hw_baseboard_id`, `sp_type`, `sp_slot` corresponds to the actual number of SPs in the rack:
```
root@[fd00:1122:3344:105::3]:32221/omicron> select distinct hw_baseboard_id, sp_type, sp_slot from inv_service_processor order by sp_type, sp_slot;
            hw_baseboard_id            | sp_type | sp_slot
---------------------------------------+---------+----------
  c6313404-bde5-42f8-a884-9de41c06843f | sled    |       8
  ede87635-3fa6-4957-b109-2309841f55f5 | sled    |       9
  0d40687a-8048-4a03-8fd5-8924478576b0 | sled    |      10
  58338408-3428-40af-8e91-522583eabbde | sled    |      11
  60ab92e2-5fdc-45f7-bbe4-3ccaef6c076c | sled    |      12
  199c809a-2cde-4864-b1b4-31066e92fccd | sled    |      13
  a69500c7-8753-4de6-8498-2c213363aba4 | sled    |      14
  04cf6209-2071-46ff-a10e-026751ada85a | sled    |      16
  7a7ecd12-f4c1-4fa6-970e-45e79c972854 | sled    |      17
  5bc79ceb-dd38-4bfc-b44f-04bf7f6a2d6a | sled    |      19
  88093ac8-f4cd-4b25-8f36-c3713aa52d93 | sled    |      21
  9587940e-760b-4fbb-ae4d-c8c711c9f561 | sled    |      23
  b352e992-2390-441d-b8eb-58005bf179c3 | sled    |      25
  0c78a1ed-bcf5-44e5-aab0-a6c3998fdc13 | sled    |      27
  37c56f3d-29e5-488a-a25a-9387692929fd | switch  |       0
  3c670d32-05a0-40e4-9981-fc655295abe8 | switch  |       1
  3b63ab8a-302e-468f-aa2a-9b7db008a53b | power   |       0
(17 rows)
```

**Comment by askfongjojo** [comment]
@davepacheco , @smklein - Are you familiar with this part of PR #4545 to know why we'd hit the 60 row constraint?

**Comment by smklein** [comment]
- Looks like the `uninitialized_sled_list` API (added in https://github.com/oxidecomputer/omicron/pull/4504) isn't paginated, but should be? However, I'm not sure that would resolve the problem you're actually encountering immediately.
- That limit comes from `db-queries/src/db/datastore/inventory.rs` (from `inventory_collection_read_all_or_nothing`), but the actual limit of `50` records comes from `uninitialized_sled_list` in `nexus/src/app/rack.rs`. Maybe there was an assumption that "~one SP per sled, plus a few extra" would be sufficient for the single-rack case?

I think these assumptions are brittle, and probably worth fixing (via pagination + proper lookup), but I think they were made intentionally given time constraints.

@davepacheco  -- any idea why the `inv_service_processor` has ~4 entries per `hw_baseboard_id`?

Asking another way: Should these four rows be a single row / single `inv_collection_id`?

```
root@[fd00:1122:3344:105::3]:32221/omicron> select * from inv_service_processor order by sp_type, sp_slot;
           inv_collection_id           |           hw_baseboard_id            |        time_collected         |                source                | sp_type | sp_slot | baseboard_revision | hubris_archive_id | power_state
---------------------------------------+--------------------------------------+-------------------------------+--------------------------------------+---------+---------+--------------------+-------------------+--------------
  a15e1a80-d9aa-4679-ba21-896ba86d9de8 | c6313404-bde5-42f8-a884-9de41c06843f | 2023-11-30 18:51:35.736285+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |       8 |                  6 | 62882ff9ad414aae  | A0
  5247e759-2478-4be8-b8ed-795e1d2ed15c | c6313404-bde5-42f8-a884-9de41c06843f | 2023-12-06 01:00:25.684695+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |       8 |                  6 | 62882ff9ad414aae  | A0
  68b2d1c0-248d-4f00-9a20-a3153a20c47f | c6313404-bde5-42f8-a884-9de41c06843f | 2023-12-06 00:59:43.813077+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |       8 |                  6 | 62882ff9ad414aae  | A0
  4e31b5ee-f2b9-4288-bcea-34a37a4f2a2e | c6313404-bde5-42f8-a884-9de41c06843f | 2023-12-06 00:59:43.236164+00 | http://[fd00:1122:3344:10b::2]:12225 | sled    |       8 |                  6 | 62882ff9ad414aae  | A0
  ```

**Comment by smklein** [comment]
TLDR of my last comment, if that didn't make sense:

- We're hitting the limit of 50, because I think that was a hard-coded value for a single-rack? We should fix, but I'm surprised we blew past it.
- The 4x duplication in the `inv_service_processor` seems to be the  underlying cause of this issue, dunno why that's happening. If this is unexpected, I think we should fix this first.

**Comment by smklein** [comment]
Okay, reading the source code a bit, I think each `inv_collection_id` indicates "who observed the SP" rather than "what is the observed SP". I think the assumption that "50 would be less than the count in one rack" breaks down here, without deduplciation by `hw_baseboard_id`

**Comment by askfongjojo** [comment]
I agree that the duplication is the higher-priority item to fix since the decision to defer pagination is very likely based on the <50 SP records assumption. However, it appears that each of the 3 nexus zones runs a collector background task by design?
```
root@[fd00:1122:3344:105::3]:32221/omicron> select * from inv_collection;
                   id                  |         time_started          |           time_done           |              collector
---------------------------------------+-------------------------------+-------------------------------+---------------------------------------
  47dac073-343a-49da-9c22-a65ac7f47444 | 2023-12-06 05:09:42.549829+00 | 2023-12-06 05:09:53.117561+00 | 20b100d0-84c3-4119-aa9b-0c632b0b6a3a
  8118aedc-1abe-4cab-8128-857daaf0ec03 | 2023-12-06 05:09:42.321334+00 | 2023-12-06 05:09:52.416581+00 | 65a11c18-7f59-41ac-b9e7-680627f996e7
  9b7de78c-a2c3-4d6a-9839-ca3c0a1bb543 | 2023-12-06 05:10:24.316779+00 | 2023-12-06 05:10:33.88068+00  | 2898657e-4141-4c05-851b-147bffc6bbbd
  a15e1a80-d9aa-4679-ba21-896ba86d9de8 | 2023-11-30 18:51:34.188247+00 | 2023-11-30 18:51:43.50998+00  | 2898657e-4141-4c05-851b-147bffc6bbbd
(4 rows)
```
The only thing unclear is why the nexus zone wiuth uuid `2898657e-4141-4c05-851b-147bffc6bbbd` didn't have the 11/30 collector taken down and its entries are therefore staying indefinitely without being [pruned](https://github.com/oxidecomputer/omicron/blob/b3d641a2dd3bb3f3dd68a413ddd670e3dbcc0b5c/nexus/src/app/background/inventory_collection.rs#L102-L107).

Even without the old collector left behind, the 3x SP entries would already exceed the 50-row limit in a half-rack deployment.

**Comment by jgallagher** [comment]
Everything noted here so far about the inventory system looks correct/expected to me, I think:

* `inv_service_processor` _is_ expected to have duplicate SP entries over time; each `collection_id` should map to a set of "the SPs that were observed when this collection was taken", and if the set of SPs doesn't change they'll be repeated. The overall size of this table should be limited by the inventory background task pruning old collection IDs in their entirety.
* Pruning ought to keep the number of collections to 3 generally, but can be 4 if:
    * the most recent 3 collections had at least one error
    * there exists a collection that did not have an error
 in which case we should have the 3 most recent (all with one or more errors) and the last fully-successful collection. I haven't looked at the db to see if that 11-30 collection meets this criteria, but that'd be my guess.

I'm still surprised we're hitting the 50 limit though - that's passed into `inventory_collection_read_all_or_nothing` along with a _specific_ collection ID, which should filter out all the duplicate SP entries in `inv_collection` (since we're limiting to just one of those collections).

**Comment by jgallagher** [comment]
The guess about why we have 4 collections instead of 3 is correct: the 11-30 collection had 0 errors, so we kept that one plus the three most recent (each of which had 16 errors):

```console
# ./omdb db inventory collections list
ID                                   STARTED              TOOK     NSPS NERRORS
a15e1a80-d9aa-4679-ba21-896ba86d9de8 2023-11-30T18:51:34Z 9321 ms  17   0
51cae547-31af-48c2-94c5-4b1a344a651b 2023-12-06T13:59:44Z 10209 ms 16   16
88450624-ccfe-4896-b1c7-70988598828e 2023-12-06T13:59:44Z 10701 ms 16   16
198c30af-88f9-4a02-9c35-7b20fa0dbe10 2023-12-06T14:00:26Z 9615 ms  16   16
```

**Comment by jgallagher** [comment]
:facepalm: I can reproduce this error with `omdb`:

```console
root@oxz_switch:/tmp/john# ./omdb db --fetch-limit 50 inventory collections show 198c30af-88f9-4a02-9c35-7b20fa0dbe10 
...
WARN: loading collection: found 50 items (the limit).  There may be more items that were ignored.  Consider overriding with --fetch-limit.
...
```

Comparing the output of that command with a higher fetch limit, we're missing some cabooses and some RoT pages. _Four_ of each of those is collected per sled, but with a fixed limit of 50, that means we ought to see (13 sleds + 2 sidecars + 1 psc) * 4 = 64 cabooses and 64 RoT pages, and indeed if I bump the fetch limit to 65 the warning/error goes away.

Fixing this should be easy but I'm not sure which is better:

* Bump the 50 up to something like 200 (a full rack would be 35*4 = 140 cabooses/rot pages)
* Inside `inventory_collection_read_all_or_nothing`, multiply the limit by N for queries where we expect to get N copies per sled

Happy to defer to @davepacheco if you have thoughts?

**Comment by davepacheco** [comment]
I think we should bump the "50" limit for now.  @jgallagher is taking this on.

As follow-on work, I think we should add a test so that we will notice if this limit gets too small again.  We can also consider better approaches here but it's trickier than it sounds.  I'll elaborate shortly.
